### DISCUSSION TITLE
convert seurat v5 to anndata

### DISCUSSION THREAD

**Discussion by rkulchar** [satijalab/seurat#8642] [General]
Hi -- thanks for your help. I began this question on #8635 but am still having issues. I am currently working with single cell (scRNAseq) and spatial transcriptomics (Xenium) datasets in Seurat v5 and was running into some issues when I try to export the h5 object to perform further analyses in Python. I was wondering if you may have any suggestions in this regard.

I get this error when I try this: > as.anndata(x = st, file_path = "/home/ICI", file_name = "st_anndata.h5ad")
• Checking Seurat object validity & Extracting Data
Error in as.anndata():
! main_layer must be one of counts.1, counts.2, counts.3, counts.4, data.1, data.2, data.3, and data.4
Run rlang::last_trace() to see where the error occurred.

My st structure is this:
An object of class Seurat
541 features across 280758 samples within 4 assays
Active assay: Xenium (280 features, 280 variable features)
9 layers present: counts.1, counts.2, counts.3, counts.4, data.1, data.2, data.3, data.4, scale.data
3 other assays present: BlankCodeword, ControlCodeword, ControlProbe
2 dimensional reductions calculated: pca, umap
4 spatial fields of view present: fov fov.2 fov.3 fov.4

**Comment by samuel-marsh** [comment]
Hi @rkulchar,

Moving answer to your other question here in terms of how to transfer spatial assays to anndata.  The best advice I can give is to try and understand the how spatial data is stored in anndata and how that is properly linked to gene expression data from spatial assays.  If you (or someone else) can figure that out then we could probably modify scCustomize's `as.anndata` fairly easily.  As I mentioned in other issue I don't have bandwidth to go through anndata structure currently but if you or another member of community understands those things then I'm happy to help with PR for scCustomize.

Best,
Sam

**Comment by samuel-marsh** [comment]
Also as this isn't error/bug with Seurat (and SeuratDisk is no longer actively supported) I'm going to move this issue to discussion section.

Best,
Sam

**Comment by rkulchar** [comment]
Thanks, Sam. I will try to figure this out...

**Reply by samuel-marsh** [reply]
@rkulchar I've got some ideas on how to implement.  I will start working on testing them this afternoon.  I will post updates both here and on release notes issue of scCustomize (https://github.com/samuel-marsh/scCustomize/issues/167). 
Best,
Sam

**Reply by rkulchar** [reply]
You are amazing! If we can figure this out, I can move to spatial analyses in Python and would love to include this and your help in a paper

**Reply by rkulchar** [reply]
let me know when and i can test in real time if i am able to save in R...also, happy to meet virtually today too

**Reply by rkulchar** [reply]
hi! just checking in @samuel-marsh

**Reply by samuel-marsh** [reply]
Hi @rkulchar 

Sorry making progress slowly.  The last thing tripping me up is this line from their docs:

`spot_diameter_fullres` - this is the diameter of the capture area for each observation. In the case of Visium, we usually call them "spots" and this value is set to ~89

I'm just not sure where 89 is coming from.  It's not the diameter of single spot in visium nor is it the approximate center to center distance.  This is something that needs to be entered when creating anndata spatial object to render and plot correctly once transferred.  

If you have any ideas let me know.  Worst, case this could be user entered parameter but that's dangerous to get wrong and much better to be something we can pull directly from the spatial assay.

Also because he might potentially know (I hope you don't mind me tagging you here) @AustinHartman do you know where the 89 value comes from in squidpy for visium spot size?  Is there a way to arrive at this value from what is stored directly in spatial assay?

Thanks!!
Sam

**Comment by rkulchar** [comment]
No worries...was able to map the data for now by just extracting layers needed...but this still would be such a helpful tool if can do :)

**Comment by SSandraL** [comment]
I have used the latest version of scCustomize 

`> as.anndata(x = macs,  file_path = "data/",  file_name = "pbmc_anndata.h5ad",  assay = "SCT",  main_layer = "data",  other_layers = c("counts"),   transer_dimreduc = TRUE, verbose = TRUE)

• Checking Seurat object validity
• Extracting Data from SCT assay to transfer to anndata.
The following columns were removed as they contain identical values for all rows:
ℹ scDoublFinder
• Creating anndata object.
• Writing anndata file: "/Users/data/pbmc_anndata.h5ad"
AnnData object with n_obs × n_vars = 7276 × 17716
    obs: 'orig.ident', 'nCount_RNA', 'nFeature_RNA', 'percent.mt', 'cluster', 'percent.rb', 'S.Score', 'G2M.Score', 'Phase', 'CC.Difference', 'orig.ident_1', 'nCount_SCT', 'nFeature_SCT', 'SCT_snn_res.0.5', 'SCT_snn_res.0.55', 'SCT_snn_res.0.6', 'SCT_snn_res.0.65', 'SCT_snn_res.0.7', 'SCT_snn_res.0.75', 'SCT_snn_res.0.8', 'SCT_snn_res.0.85', 'SCT_snn_res.0.9', 'seurat_clusters', 'Group'
    var: 'names'
    obsm: 'X_pca', 'X_umap', 'X_integrated.dr'
    layers: 'counts_SCT'
Warning messages:
1: Adding a command log without an assay associated with it 
2: Adding a command log without an assay associated with it 
3: Adding a command log without an assay associated with it 
`
When I tried to open the it using python I get an error. 

`adata = sc.read("../data/pbmc_anndata.h5ad")

TypeError: Cannot convert numpy.ndarray to numpy.ndarray
Error raised while reading key 'orig.ident' of <class 'h5py._hl.group.Group'> from [/obs](http://localhost:8888/obs)`

orig.ident contains the mane of my samples and is a character.

**Reply by samuel-marsh** [reply]
See comment in scCustomize repo.
